Rcode

mamueller opened the pull request:
I implemented an Rcode printer. I basically copied the ccode printer along with its tests and adapted both to produce valid R code from symy expressions.  I do not know about any problems.
I have been using part of the functionality to check numerical solutions of an R package against symbolic solutions computed with sympy and then translated to R code automatically.

mamueller committed:
initial copy

mamueller committed:
Merge branch 'master' into Rcode
fetched the latest updates

mamueller committed:
changed names so that the new module is loaded and tested, although it still translates into c not R

mamueller committed:
renamed class

mamueller committed:
made the test for power expressions green in test_rcode2.py

mamueller committed:
all tests in test_rcode green

mamueller committed:
nested piecewise test now runs in test_rcode2.py

mamueller committed:
corrected some tabs and found a wrong test. There is a paranthesis
missing int piecewise expressions

mamueller committed:
added a wrong priecewise
test

mamueller committed:
Repaired the piecewise test

mamueller committed:
tried to find out how the constants are handled and introduced a lot of
tracing that will be erased later

mamueller committed:
The tests for the constants now work
also removed the verbose debugging printouts

mamueller committed:
changed the translation for the for loops

mamueller committed:
Checked a little more of the translation. The next step is concerned
with arrays (matrices)

mamueller committed:
finished the translation of the tests of the ccode to the rcode module

mamueller committed:
corrected a comment that still refered to C as the target language

mamueller committed:
Merge branch 'master' into Rcode
tried to resolve conflicts, have to wait for the tests

mamueller committed:
Adapted the Rcode printer to the current version of sympy. All tests and doctests work

mamueller committed:
Merge remote-tracking branch 'upstream/master' into Rcode

mamueller committed:
removed a superflous print statement that I had forgotten.

moorepants commented on sympy/printing/printer.py:
```diff
@@ -230,7 +230,8 @@ def order(self):
         """Returns printer's representation for expr (as a string)"""
-        return self._str(self._print(expr))
+        res=self._str(self._print(expr))
+        return res
```
What is the reason for this change?

mamueller replied:
Sorry, just a leftover of "print debugging..." I removed the print but forgot to inline it again, I will do it now along with the removal of the white space that I forgot as well..

moorepants commented on sympy/printing/printer.py:
```diff
@@ -254,10 +255,12 @@ def _print(self, expr, *args, **kwargs):
             # Unknown object, fall back to the emptyPrinter.
-            return self.emptyPrinter(expr)
+            res=self.emptyPrinter(expr)
+            return res
```
I don't see any reason for this change either.

mamueller replied:
Same as above...

moorepants commented:
This is great! Thanks so much for this addition.

A few things need to be fixed:

- Remove any trailing whitespace. You can run the test_code_quality.py file to check. Also see the Travis failures.
- Can you add documentation? For example, make sure the API docs show up like: http://docs.sympy.org/latest/modules/printing.html?highlight=ccode#sympy.printing.ccode. And also would you mind adding a bit to here: http://docs.sympy.org/latest/modules/codegen.html#code-printers-sympy-printing
- Can you run `bin/coverage_report.py` and report the results of your test coverage here?

Other than, that I think this is looking great!

mamueller committed:
Implemented the proposed changes to the comment on the pullrequest.

mamueller committed:
added the new parts to the documentation of the printing module

mamueller commented:
Coverage report: 77% Show keyboard shortcuts
Module 	statements 	missing 	excluded 	coverage
Total 	118475 	26949 	742 	77%
sympy/__init__.py 	51 	51 	0 	0%
sympy/abc.py 	40 	0 	0 	100%
sympy/assumptions/__init__.py 	3 	3 	0 	0%
sympy/assumptions/ask.py 	224 	127 	0 	43%
sympy/assumptions/ask_generated.py 	7 	5 	0 	29%
sympy/assumptions/assume.py 	85 	37 	0 	56%
sympy/assumptions/handlers/__init__.py 	1 	0 	0 	100%
sympy/assumptions/handlers/calculus.py 	61 	21 	0 	66%
sympy/assumptions/handlers/common.py 	71 	7 	0 	90%
sympy/assumptions/handlers/matrices.py 	260 	26 	0 	90%
sympy/assumptions/handlers/ntheory.py 	143 	9 	0 	94%
sympy/assumptions/handlers/order.py 	215 	15 	0 	93%
sympy/assumptions/handlers/sets.py 	342 	20 	0 	94%
sympy/assumptions/refine.py 	106 	14 	0 	87%
sympy/assumptions/satask.py 	48 	1 	0 	98%
sympy/assumptions/sathandlers.py 	122 	6 	0 	95%
sympy/calculus/__init__.py 	4 	4 	0 	0%
sympy/calculus/euler.py 	28 	4 	0 	86%
sympy/calculus/finite_diff.py 	75 	14 	0 	81%
sympy/calculus/singularities.py 	59 	12 	6 	80%
sympy/calculus/util.py 	499 	119 	3 	76%
sympy/categories/__init__.py 	2 	0 	0 	100%
sympy/categories/baseclasses.py 	168 	0 	2 	100%
sympy/categories/diagram_drawing.py 	861 	29 	0 	97%
sympy/codegen/__init__.py 	1 	1 	0 	0%
sympy/codegen/ast.py 	109 	41 	3 	62%
sympy/combinatorics/__init__.py 	10 	0 	0 	100%
sympy/combinatorics/fp_groups.py 	735 	73 	0 	90%
sympy/combinatorics/free_groups.py 	421 	44 	0 	90%
sympy/combinatorics/generators.py 	154 	1 	0 	99%
sympy/combinatorics/graycode.py 	92 	5 	0 	95%
sympy/combinatorics/group_constructs.py 	31 	0 	0 	100%
sympy/combinatorics/named_groups.py 	111 	1 	0 	99%
sympy/combinatorics/partitions.py 	230 	1 	0 	99%
sympy/combinatorics/perm_groups.py 	1139 	77 	1 	93%
sympy/combinatorics/permutations.py 	733 	31 	2 	96%
sympy/combinatorics/polyhedron.py 	97 	0 	0 	100%
sympy/combinatorics/prufer.py 	137 	4 	0 	97%
sympy/combinatorics/subsets.py 	140 	6 	0 	96%
sympy/combinatorics/tensor_can.py 	413 	16 	0 	96%
sympy/combinatorics/testutil.py 	137 	3 	0 	98%
sympy/combinatorics/util.py 	139 	9 	0 	94%
sympy/concrete/__init__.py 	2 	2 	0 	0%
sympy/concrete/delta.py 	138 	7 	0 	95%
sympy/concrete/expr_with_intlimits.py 	61 	17 	0 	72%
sympy/concrete/expr_with_limits.py 	214 	40 	0 	81%
sympy/concrete/gosper.py 	82 	15 	0 	82%
sympy/concrete/guess.py 	169 	29 	0 	83%
sympy/concrete/products.py 	144 	38 	1 	74%
sympy/concrete/summations.py 	496 	79 	2 	84%
sympy/conftest.py 	50 	50 	0 	0%
sympy/core/__init__.py 	21 	21 	0 	0%
sympy/core/add.py 	595 	76 	0 	87%
sympy/core/alphabets.py 	2 	2 	0 	0%
sympy/core/assumptions.py 	100 	20 	0 	80%
sympy/core/backend.py 	9 	3 	0 	67%
sympy/core/basic.py 	601 	151 	0 	75%
sympy/core/cache.py 	95 	76 	0 	20%
sympy/core/compatibility.py 	290 	208 	0 	28%
sympy/core/containers.py 	110 	38 	1 	65%
sympy/core/core.py 	39 	13 	0 	67%
sympy/core/coreerrors.py 	3 	3 	0 	0%
sympy/core/decorators.py 	45 	12 	0 	73%
sympy/core/evalf.py 	852 	143 	12 	83%
sympy/core/evaluate.py 	12 	6 	0 	50%
sympy/core/expr.py 	1417 	225 	5 	84%
sympy/core/exprtools.py 	767 	101 	0 	87%
sympy/core/facts.py 	204 	41 	0 	80%
sympy/core/function.py 	841 	194 	3 	77%
sympy/core/logic.py 	177 	49 	0 	72%
sympy/core/mod.py 	94 	8 	0 	91%
sympy/core/mul.py 	1015 	140 	0 	86%
sympy/core/multidimensional.py 	62 	49 	0 	21%
sympy/core/numbers.py 	2033 	697 	6 	66%
sympy/core/operations.py 	216 	37 	0 	83%
sympy/core/power.py 	973 	96 	2 	90%
sympy/core/relational.py 	238 	76 	1 	68%
sympy/core/rules.py 	15 	5 	0 	67%
sympy/core/singleton.py 	35 	19 	0 	46%
sympy/core/symbol.py 	229 	69 	0 	70%
sympy/core/sympify.py 	145 	35 	0 	76%
sympy/core/trace.py 	75 	2 	0 	97%
sympy/crypto/__init__.py 	1 	0 	0 	100%
sympy/crypto/crypto.py 	370 	2 	0 	99%
sympy/deprecated/__init__.py 	1 	1 	0 	0%
sympy/deprecated/class_registry.py 	16 	16 	0 	0%
sympy/diffgeom/__init__.py 	1 	0 	0 	100%
sympy/diffgeom/diffgeom.py 	432 	19 	4 	96%
sympy/diffgeom/rn.py 	39 	0 	0 	100%
sympy/external/__init__.py 	1 	1 	0 	0%
sympy/external/importtools.py 	52 	35 	0 	33%
sympy/functions/__init__.py 	24 	24 	0 	0%
sympy/functions/combinatorial/__init__.py 	2 	2 	0 	0%
sympy/functions/combinatorial/factorials.py 	383 	98 	0 	74%
sympy/functions/combinatorial/numbers.py 	535 	120 	0 	78%
sympy/functions/elementary/__init__.py 	6 	6 	0 	0%
sympy/functions/elementary/complexes.py 	605 	144 	0 	76%
sympy/functions/elementary/exponential.py 	539 	97 	0 	82%
sympy/functions/elementary/hyperbolic.py 	741 	197 	0 	73%
sympy/functions/elementary/integers.py 	145 	31 	0 	79%
sympy/functions/elementary/miscellaneous.py 	205 	66 	0 	68%
sympy/functions/elementary/piecewise.py 	352 	39 	3 	89%
sympy/functions/elementary/trigonometric.py 	1488 	366 	1 	75%
sympy/functions/special/__init__.py 	10 	10 	0 	0%
sympy/functions/special/bessel.py 	600 	229 	4 	62%
sympy/functions/special/beta_functions.py 	22 	13 	0 	41%
sympy/functions/special/bsplines.py 	48 	11 	0 	77%
sympy/functions/special/delta_functions.py 	134 	45 	0 	66%
sympy/functions/special/elliptic_integrals.py 	179 	41 	0 	77%
sympy/functions/special/error_functions.py 	796 	305 	0 	62%
sympy/functions/special/gamma_functions.py 	402 	84 	0 	79%
sympy/functions/special/hyper.py 	421 	154 	5 	63%
sympy/functions/special/mathieu_functions.py 	54 	28 	0 	48%
sympy/functions/special/polynomials.py 	354 	170 	0 	52%
sympy/functions/special/singularity_functions.py 	62 	16 	0 	74%
sympy/functions/special/spherical_harmonics.py 	84 	41 	0 	51%
sympy/functions/special/tensor_functions.py 	88 	32 	0 	64%
sympy/functions/special/zeta_functions.py 	171 	35 	0 	80%
sympy/galgebra.py 	1 	1 	0 	0%
sympy/geometry/__init__.py 	9 	9 	0 	0%
sympy/geometry/curve.py 	65 	21 	0 	68%
sympy/geometry/ellipse.py 	374 	242 	5 	35%
sympy/geometry/entity.py 	222 	117 	8 	47%
sympy/geometry/exceptions.py 	3 	3 	0 	0%
sympy/geometry/line.py 	532 	167 	2 	69%
sympy/geometry/parabola.py 	57 	19 	1 	67%
sympy/geometry/plane.py 	270 	208 	1 	23%
sympy/geometry/point.py 	317 	105 	0 	67%
sympy/geometry/polygon.py 	590 	365 	1 	38%
sympy/geometry/util.py 	258 	78 	1 	70%
sympy/holonomic/__init__.py 	2 	0 	0 	100%
sympy/holonomic/holonomic.py 	1513 	197 	7 	87%
sympy/holonomic/holonomicerrors.py 	34 	17 	1 	50%
sympy/holonomic/linearsolver.py 	54 	6 	0 	89%
sympy/holonomic/numerical.py 	69 	2 	0 	97%
sympy/holonomic/recurrence.py 	178 	53 	0 	70%
sympy/integrals/__init__.py 	3 	3 	0 	0%
sympy/integrals/deltafunctions.py 	73 	1 	0 	99%
sympy/integrals/heurisch.py 	357 	15 	0 	96%
sympy/integrals/integrals.py 	521 	99 	4 	81%
sympy/integrals/manualintegrate.py 	696 	133 	0 	81%
sympy/integrals/meijerint.py 	1113 	134 	1 	88%
sympy/integrals/meijerint_doc.py 	20 	20 	0 	0%
sympy/integrals/prde.py 	371 	54 	12 	85%
sympy/integrals/quadrature.py 	101 	1 	0 	99%
sympy/integrals/rationaltools.py 	149 	2 	0 	99%
sympy/integrals/rde.py 	361 	74 	9 	80%
sympy/integrals/risch.py 	685 	17 	7 	98%
sympy/integrals/singularityfunctions.py 	20 	5 	0 	75%
sympy/integrals/transforms.py 	750 	229 	7 	69%
sympy/integrals/trigonometry.py 	87 	18 	0 	79%
sympy/interactive/__init__.py 	2 	2 	0 	0%
sympy/interactive/ipythonprinting.py 	8 	8 	0 	0%
sympy/interactive/printing.py 	187 	84 	0 	55%
sympy/interactive/session.py 	176 	128 	0 	27%
sympy/liealgebras/__init__.py 	1 	0 	0 	100%
sympy/liealgebras/cartan_matrix.py 	3 	0 	0 	100%
sympy/liealgebras/cartan_type.py 	48 	2 	0 	96%
sympy/liealgebras/dynkin_diagram.py 	3 	0 	0 	100%
sympy/liealgebras/root_system.py 	52 	2 	0 	96%
sympy/liealgebras/type_a.py 	55 	1 	0 	98%
sympy/liealgebras/type_b.py 	67 	2 	0 	97%
sympy/liealgebras/type_c.py 	67 	2 	0 	97%
sympy/liealgebras/type_d.py 	68 	6 	0 	91%
sympy/liealgebras/type_e.py 	180 	94 	0 	48%
sympy/liealgebras/type_f.py 	68 	2 	0 	97%
sympy/liealgebras/type_g.py 	26 	2 	0 	92%
sympy/liealgebras/weyl_group.py 	228 	67 	0 	71%
sympy/logic/__init__.py 	2 	2 	0 	0%
sympy/logic/algorithms/__init__.py 	0 	0 	0 	100%
sympy/logic/algorithms/dpll.py 	158 	23 	0 	85%
sympy/logic/algorithms/dpll2.py 	217 	21 	2 	90%
sympy/logic/boolalg.py 	651 	154 	4 	76%
sympy/logic/inference.py 	75 	24 	4 	68%
sympy/logic/utilities/__init__.py 	1 	0 	0 	100%
sympy/logic/utilities/dimacs.py 	34 	3 	0 	91%
sympy/matrices/__init__.py 	9 	9 	0 	0%
sympy/matrices/dense.py 	461 	128 	24 	72%
sympy/matrices/densearith.py 	26 	0 	0 	100%
sympy/matrices/densesolve.py 	109 	18 	0 	83%
sympy/matrices/densetools.py 	50 	4 	0 	92%
sympy/matrices/expressions/__init__.py 	15 	15 	0 	0%
sympy/matrices/expressions/adjoint.py 	27 	16 	0 	41%
sympy/matrices/expressions/blockmatrix.py 	235 	84 	1 	64%
sympy/matrices/expressions/determinant.py 	31 	15 	0 	52%
sympy/matrices/expressions/diagonal.py 	13 	7 	0 	46%
sympy/matrices/expressions/dotproduct.py 	28 	8 	0 	71%
sympy/matrices/expressions/factorizations.py 	34 	1 	0 	97%
sympy/matrices/expressions/fourier.py 	17 	3 	0 	82%
sympy/matrices/expressions/funcmatrix.py 	20 	12 	0 	40%
sympy/matrices/expressions/hadamard.py 	37 	14 	0 	62%
sympy/matrices/expressions/inverse.py 	39 	26 	0 	33%
sympy/matrices/expressions/matadd.py 	56 	24 	0 	57%
sympy/matrices/expressions/matexpr.py 	291 	147 	4 	49%
sympy/matrices/expressions/matmul.py 	164 	39 	1 	76%
sympy/matrices/expressions/matpow.py 	50 	16 	1 	68%
sympy/matrices/expressions/slice.py 	63 	15 	0 	76%
sympy/matrices/expressions/trace.py 	33 	13 	0 	61%
sympy/matrices/expressions/transpose.py 	42 	22 	0 	48%
sympy/matrices/immutable.py 	89 	59 	0 	34%
sympy/matrices/matrices.py 	1439 	266 	17 	82%
sympy/matrices/sparse.py 	602 	138 	3 	77%
sympy/matrices/sparsetools.py 	19 	0 	0 	100%
sympy/ntheory/__init__.py 	8 	8 	0 	0%
sympy/ntheory/bbp_pi.py 	41 	0 	0 	100%
sympy/ntheory/continued_fraction.py 	65 	5 	0 	92%
sympy/ntheory/egyptian_fraction.py 	73 	16 	0 	78%
sympy/ntheory/factor_.py 	664 	74 	0 	89%
sympy/ntheory/generate.py 	306 	31 	0 	90%
sympy/ntheory/modular.py 	74 	2 	0 	97%
sympy/ntheory/multinomial.py 	113 	38 	0 	66%
sympy/ntheory/partitions_.py 	114 	14 	0 	88%
sympy/ntheory/primetest.py 	231 	85 	0 	63%
sympy/ntheory/residue_ntheory.py 	649 	52 	1 	92%
sympy/parsing/__init__.py 	0 	0 	0 	100%
sympy/parsing/ast_parser.py 	41 	30 	0 	27%
sympy/parsing/mathematica.py 	22 	0 	0 	100%
sympy/parsing/maxima.py 	36 	0 	0 	100%
sympy/parsing/sympy_parser.py 	432 	4 	0 	99%
sympy/parsing/sympy_tokenize.py 	251 	86 	0 	66%
sympy/physics/__init__.py 	2 	0 	0 	100%
sympy/physics/continuum_mechanics/__init__.py 	1 	0 	0 	100%
sympy/physics/continuum_mechanics/beam.py 	122 	2 	0 	98%
sympy/physics/gaussopt.py 	3 	3 	0 	0%
sympy/physics/hep/__init__.py 	0 	0 	0 	100%
sympy/physics/hep/gamma_matrices.py 	387 	31 	0 	92%
sympy/physics/hydrogen.py 	27 	0 	0 	100%
sympy/physics/matrices.py 	51 	2 	0 	96%
sympy/physics/mechanics/__init__.py 	28 	0 	0 	100%
sympy/physics/mechanics/body.py 	42 	0 	0 	100%
sympy/physics/mechanics/functions.py 	154 	24 	0 	84%
sympy/physics/mechanics/kane.py 	413 	88 	0 	79%
sympy/physics/mechanics/lagrange.py 	175 	24 	0 	86%
sympy/physics/mechanics/linearize.py 	229 	29 	0 	87%
sympy/physics/mechanics/models.py 	88 	2 	0 	98%
sympy/physics/mechanics/particle.py 	51 	13 	0 	75%
sympy/physics/mechanics/rigidbody.py 	92 	24 	0 	74%
sympy/physics/mechanics/system.py 	135 	0 	0 	100%
sympy/physics/optics/__init__.py 	13 	0 	0 	100%
sympy/physics/optics/gaussopt.py 	129 	7 	2 	95%
sympy/physics/optics/medium.py 	45 	6 	0 	87%
sympy/physics/optics/utils.py 	177 	25 	0 	86%
sympy/physics/optics/waves.py 	63 	8 	1 	87%
sympy/physics/paulialgebra.py 	41 	1 	0 	98%
sympy/physics/pring.py 	11 	1 	0 	91%
sympy/physics/qho_1d.py 	13 	0 	0 	100%
sympy/physics/quantum/__init__.py 	34 	0 	0 	100%
sympy/physics/quantum/anticommutator.py 	56 	1 	0 	98%
sympy/physics/quantum/boson.py 	112 	22 	0 	80%
sympy/physics/quantum/cartesian.py 	154 	2 	0 	99%
sympy/physics/quantum/cg.py 	379 	39 	1 	90%
sympy/physics/quantum/circuitplot.py 	178 	137 	0 	23%
sympy/physics/quantum/circuitutils.py 	162 	37 	0 	77%
sympy/physics/quantum/commutator.py 	102 	0 	0 	100%
sympy/physics/quantum/constants.py 	26 	0 	0 	100%
sympy/physics/quantum/dagger.py 	15 	3 	0 	80%
sympy/physics/quantum/density.py 	81 	17 	0 	79%
sympy/physics/quantum/fermion.py 	81 	20 	0 	75%
sympy/physics/quantum/gate.py 	510 	49 	6 	90%
sympy/physics/quantum/grover.py 	78 	8 	0 	90%
sympy/physics/quantum/hilbert.py 	285 	19 	1 	93%
sympy/physics/quantum/identitysearch.py 	197 	40 	0 	80%
sympy/physics/quantum/innerproduct.py 	54 	5 	0 	91%
sympy/physics/quantum/matrixcache.py 	51 	4 	1 	92%
sympy/physics/quantum/matrixutils.py 	190 	102 	2 	46%
sympy/physics/quantum/operator.py 	185 	23 	3 	88%
sympy/physics/quantum/operatorordering.py 	141 	43 	0 	70%
sympy/physics/quantum/operatorset.py 	83 	18 	3 	78%
sympy/physics/quantum/pauli.py 	314 	142 	6 	55%
sympy/physics/quantum/piab.py 	35 	0 	0 	100%
sympy/physics/quantum/qapply.py 	80 	0 	0 	100%
sympy/physics/quantum/qasm.py 	121 	8 	0 	93%
sympy/physics/quantum/qexpr.py 	151 	16 	3 	89%
sympy/physics/quantum/qft.py 	110 	14 	1 	87%
sympy/physics/quantum/qubit.py 	252 	42 	4 	83%
sympy/physics/quantum/represent.py 	196 	24 	6 	88%
sympy/physics/quantum/sho1d.py 	203 	47 	4 	77%
sympy/physics/quantum/shor.py 	96 	63 	1 	34%
sympy/physics/quantum/spin.py 	947 	77 	3 	92%
sympy/physics/quantum/state.py 	228 	9 	6 	96%
sympy/physics/quantum/tensorproduct.py 	163 	36 	0 	78%
sympy/physics/secondquant.py 	1040 	153 	3 	85%
sympy/physics/sho.py 	10 	0 	0 	100%
sympy/physics/units.py 	150 	0 	0 	100%
sympy/physics/unitsystems/__init__.py 	4 	0 	0 	100%
sympy/physics/unitsystems/dimensions.py 	191 	7 	1 	96%
sympy/physics/unitsystems/prefixes.py 	46 	7 	0 	85%
sympy/physics/unitsystems/quantities.py 	66 	13 	1 	80%
sympy/physics/unitsystems/simplifiers.py 	60 	14 	0 	77%
sympy/physics/unitsystems/systems/__init__.py 	3 	0 	0 	100%
sympy/physics/unitsystems/systems/mks.py 	40 	0 	0 	100%
sympy/physics/unitsystems/systems/mksa.py 	34 	0 	0 	100%
sympy/physics/unitsystems/systems/natural.py 	23 	0 	0 	100%
sympy/physics/unitsystems/units.py 	172 	8 	0 	95%
sympy/physics/vector/__init__.py 	22 	0 	0 	100%
sympy/physics/vector/dyadic.py 	213 	48 	0 	77%
sympy/physics/vector/fieldfunctions.py 	75 	0 	0 	100%
sympy/physics/vector/frame.py 	302 	47 	1 	84%
sympy/physics/vector/functions.py 	231 	46 	0 	80%
sympy/physics/vector/point.py 	130 	4 	0 	97%
sympy/physics/vector/printing.py 	176 	78 	0 	56%
sympy/physics/vector/vector.py 	294 	49 	0 	83%
sympy/physics/wigner.py 	143 	9 	0 	94%
sympy/plotting/__init__.py 	5 	5 	0 	0%
sympy/plotting/experimental_lambdify.py 	238 	127 	0 	47%
sympy/plotting/intervalmath/__init__.py 	2 	2 	0 	0%
sympy/plotting/intervalmath/interval_arithmetic.py 	286 	117 	0 	59%
sympy/plotting/intervalmath/lib_interval.py 	275 	275 	11 	0%
sympy/plotting/plot.py 	667 	576 	0 	14%
sympy/plotting/plot_implicit.py 	171 	120 	2 	30%
sympy/plotting/textplot.py 	53 	11 	0 	79%
sympy/polys/__init__.py 	52 	52 	0 	0%
sympy/polys/agca/__init__.py 	1 	0 	0 	100%
sympy/polys/agca/homomorphisms.py 	178 	0 	10 	100%
sympy/polys/agca/ideals.py 	109 	1 	21 	99%
sympy/polys/agca/modules.py 	420 	18 	16 	96%
sympy/polys/compatibility.py 	951 	755 	1 	21%
sympy/polys/constructor.py 	161 	13 	0 	92%
sympy/polys/densearith.py 	580 	87 	0 	85%
sympy/polys/densebasic.py 	570 	97 	0 	83%
sympy/polys/densetools.py 	421 	58 	0 	86%
sympy/polys/dispersion.py 	46 	3 	0 	93%
sympy/polys/distributedmodules.py 	212 	8 	0 	96%
sympy/polys/domains/__init__.py 	67 	67 	0 	0%
sympy/polys/domains/algebraicfield.py 	74 	41 	0 	45%
sympy/polys/domains/characteristiczero.py 	8 	8 	0 	0%
sympy/polys/domains/complexfield.py 	68 	50 	0 	26%
sympy/polys/domains/compositedomain.py 	12 	9 	0 	25%
sympy/polys/domains/domain.py 	293 	152 	18 	48%
sympy/polys/domains/domainelement.py 	5 	5 	1 	0%
sympy/polys/domains/expressiondomain.py 	135 	82 	0 	39%
sympy/polys/domains/field.py 	39 	20 	0 	49%
sympy/polys/domains/finitefield.py 	63 	42 	0 	33%
sympy/polys/domains/fractionfield.py 	79 	45 	0 	43%
sympy/polys/domains/gmpyfinitefield.py 	9 	9 	0 	0%
sympy/polys/domains/gmpyintegerring.py 	50 	50 	0 	0%
sympy/polys/domains/gmpyrationalfield.py 	50 	50 	0 	0%
sympy/polys/domains/groundtypes.py 	32 	30 	0 	6%
sympy/polys/domains/integerring.py 	23 	19 	0 	17%
sympy/polys/domains/modularinteger.py 	133 	58 	0 	56%
sympy/polys/domains/mpelements.py 	119 	47 	0 	61%
sympy/polys/domains/old_fractionfield.py 	92 	30 	2 	67%
sympy/polys/domains/old_polynomialring.py 	165 	16 	3 	90%
sympy/polys/domains/polynomialring.py 	81 	42 	0 	48%
sympy/polys/domains/pythonfinitefield.py 	9 	8 	0 	11%
sympy/polys/domains/pythonintegerring.py 	48 	29 	0 	40%
sympy/polys/domains/pythonrational.py 	161 	57 	0 	65%
sympy/polys/domains/pythonrationalfield.py 	41 	24 	0 	41%
sympy/polys/domains/quotientring.py 	102 	6 	2 	94%
sympy/polys/domains/rationalfield.py 	18 	14 	0 	22%
sympy/polys/domains/realfield.py 	71 	42 	0 	41%
sympy/polys/domains/ring.py 	51 	28 	1 	45%
sympy/polys/domains/simpledomain.py 	8 	7 	0 	12%
sympy/polys/euclidtools.py 	671 	106 	0 	84%
sympy/polys/factortools.py 	685 	54 	19 	92%
sympy/polys/fglmtools.py 	81 	10 	0 	88%
sympy/polys/fields.py 	401 	112 	3 	72%
sympy/polys/galoistools.py 	720 	92 	0 	87%
sympy/polys/groebnertools.py 	378 	85 	0 	78%
sympy/polys/heuristicgcd.py 	65 	7 	0 	89%
sympy/polys/modulargcd.py 	800 	8 	0 	99%
sympy/polys/monomials.py 	206 	66 	0 	68%
sympy/polys/numberfields.py 	613 	84 	10 	86%
sympy/polys/orderings.py 	124 	64 	1 	48%
sympy/polys/orthopolys.py 	168 	34 	0 	80%
sympy/polys/partfrac.py 	165 	24 	2 	85%
sympy/polys/polyclasses.py 	1007 	407 	0 	60%
sympy/polys/polyconfig.py 	28 	18 	0 	36%
sympy/polys/polyerrors.py 	108 	88 	14 	19%
sympy/polys/polyfuncs.py 	141 	21 	0 	85%
sympy/polys/polyoptions.py 	390 	205 	1 	47%
sympy/polys/polyquinticconst.py 	108 	22 	0 	80%
sympy/polys/polyroots.py 	631 	82 	2 	87%
sympy/polys/polytools.py 	2210 	424 	151 	81%
sympy/polys/polyutils.py 	264 	39 	4 	85%
sympy/polys/rationaltools.py 	24 	8 	0 	67%
sympy/polys/ring_series.py 	958 	193 	12 	80%
sympy/polys/rings.py 	1375 	294 	2 	79%
sympy/polys/rootisolation.py 	999 	135 	4 	86%
sympy/polys/rootoftools.py 	544 	102 	3 	81%
sympy/polys/solvers.py 	38 	1 	0 	97%
sympy/polys/specialpolys.py 	179 	49 	0 	73%
sympy/polys/sqfreetools.py 	168 	29 	2 	83%
sympy/polys/subresultants_qq_zz.py 	878 	113 	0 	87%
sympy/printing/__init__.py 	17 	17 	0 	0%
sympy/printing/ccode.py 	148 	42 	1 	72%
sympy/printing/codeprinter.py 	264 	90 	7 	66%
sympy/printing/conventions.py 	42 	7 	0 	83%
sympy/printing/defaults.py 	6 	4 	0 	33%
sympy/printing/dot.py 	56 	2 	0 	96%
sympy/printing/fcode.py 	245 	49 	2 	80%
sympy/printing/gtk.py 	11 	7 	0 	36%
sympy/printing/jscode.py 	109 	39 	0 	64%
sympy/printing/julia.py 	258 	72 	0 	72%
sympy/printing/lambdarepr.py 	146 	106 	0 	27%
sympy/printing/latex.py 	1254 	340 	0 	73%
sympy/printing/llvmjitcode.py 	259 	214 	1 	17%
sympy/printing/mathematica.py 	56 	26 	0 	54%
sympy/printing/mathml.py 	311 	102 	0 	67%
sympy/printing/octave.py 	265 	74 	0 	72%
sympy/printing/precedence.py 	57 	17 	0 	70%
sympy/printing/pretty/__init__.py 	2 	2 	0 	0%
sympy/printing/pretty/pretty.py 	1470 	343 	2 	77%
sympy/printing/pretty/pretty_symbology.py 	201 	95 	0 	53%
sympy/printing/pretty/stringpict.py 	250 	94 	1 	62%
sympy/printing/preview.py 	140 	108 	0 	23%
sympy/printing/printer.py 	49 	23 	0 	53%
sympy/printing/python.py 	57 	15 	0 	74%
sympy/printing/rcode.py 	151 	49 	1 	68%
sympy/printing/repr.py 	123 	54 	0 	56%
sympy/printing/str.py 	499 	182 	0 	64%
sympy/printing/tableform.py 	178 	21 	0 	88%
sympy/printing/theanocode.py 	126 	90 	1 	29%
sympy/printing/tree.py 	37 	37 	0 	0%
sympy/release.py 	1 	1 	0 	0%
sympy/sandbox/__init__.py 	0 	0 	0 	100%
sympy/sandbox/indexed_integrals.py 	30 	3 	0 	90%
sympy/series/__init__.py 	12 	12 	0 	0%
sympy/series/acceleration.py 	17 	0 	0 	100%
sympy/series/approximants.py 	47 	8 	0 	83%
sympy/series/formal.py 	606 	92 	1 	85%
sympy/series/fourier.py 	161 	41 	0 	75%
sympy/series/gruntz.py 	307 	60 	6 	80%
sympy/series/kauers.py 	11 	0 	0 	100%
sympy/series/limits.py 	96 	16 	2 	83%
sympy/series/limitseq.py 	83 	16 	0 	81%
sympy/series/order.py 	254 	38 	5 	85%
sympy/series/residues.py 	29 	4 	2 	86%
sympy/series/sequences.py 	373 	99 	6 	73%
sympy/series/series.py 	5 	3 	0 	40%
sympy/series/series_class.py 	43 	21 	5 	51%
sympy/sets/__init__.py 	4 	4 	0 	0%
sympy/sets/conditionset.py 	30 	12 	0 	60%
sympy/sets/contains.py 	13 	7 	0 	46%
sympy/sets/fancysets.py 	711 	122 	8 	83%
sympy/sets/sets.py 	889 	235 	9 	74%
sympy/simplify/__init__.py 	12 	12 	0 	0%
sympy/simplify/combsimp.py 	300 	16 	0 	95%
sympy/simplify/cse_main.py 	324 	27 	0 	92%
sympy/simplify/cse_opts.py 	24 	7 	0 	71%
sympy/simplify/epathtools.py 	146 	16 	1 	89%
sympy/simplify/fu.py 	897 	80 	2 	91%
sympy/simplify/hyperexpand.py 	1221 	185 	4 	85%
sympy/simplify/hyperexpand_doc.py 	9 	9 	0 	0%
sympy/simplify/powsimp.py 	334 	29 	0 	91%
sympy/simplify/radsimp.py 	442 	59 	3 	87%
sympy/simplify/ratsimp.py 	101 	14 	0 	86%
sympy/simplify/simplify.py 	613 	72 	0 	88%
sympy/simplify/sqrtdenest.py 	311 	34 	0 	89%
sympy/simplify/traversaltools.py 	14 	3 	0 	79%
sympy/simplify/trigsimp.py 	487 	277 	0 	43%
sympy/solvers/__init__.py 	10 	10 	0 	0%
sympy/solvers/bivariate.py 	178 	21 	3 	88%
sympy/solvers/decompogen.py 	29 	4 	0 	86%
sympy/solvers/deutils.py 	90 	19 	1 	79%
sympy/solvers/diophantine.py 	1297 	119 	4 	91%
sympy/solvers/inequalities.py 	324 	35 	7 	89%
sympy/solvers/ode.py 	3356 	874 	21 	74%
sympy/solvers/pde.py 	329 	73 	5 	78%
sympy/solvers/polysys.py 	125 	15 	2 	88%
sympy/solvers/recurr.py 	369 	38 	0 	90%
sympy/solvers/solvers.py 	1517 	199 	15 	87%
sympy/solvers/solveset.py 	819 	74 	3 	91%
sympy/stats/__init__.py 	16 	0 	0 	100%
sympy/stats/crv.py 	225 	14 	6 	94%
sympy/stats/crv_types.py 	358 	10 	0 	97%
sympy/stats/drv.py 	64 	19 	3 	70%
sympy/stats/drv_types.py 	28 	2 	0 	93%
sympy/stats/error_prop.py 	36 	1 	0 	97%
sympy/stats/frv.py 	183 	13 	1 	93%
sympy/stats/frv_types.py 	101 	3 	0 	97%
sympy/stats/rv.py 	392 	36 	9 	91%
sympy/stats/rv_interface.py 	25 	0 	0 	100%
sympy/stats/symbolic_probability.py 	185 	14 	1 	92%
sympy/strategies/__init__.py 	7 	7 	0 	0%
sympy/strategies/branch/__init__.py 	3 	3 	0 	0%
sympy/strategies/branch/core.py 	83 	15 	0 	82%
sympy/strategies/branch/tools.py 	5 	4 	0 	20%
sympy/strategies/branch/traverse.py 	16 	6 	0 	62%
sympy/strategies/core.py 	76 	15 	0 	80%
sympy/strategies/rl.py 	56 	11 	0 	80%
sympy/strategies/tools.py 	12 	8 	0 	33%
sympy/strategies/traverse.py 	19 	8 	0 	58%
sympy/strategies/tree.py 	18 	9 	0 	50%
sympy/strategies/util.py 	9 	7 	0 	22%
sympy/tensor/__init__.py 	3 	3 	0 	0%
sympy/tensor/array/__init__.py 	5 	5 	0 	0%
sympy/tensor/array/arrayop.py 	97 	15 	0 	85%
sympy/tensor/array/dense_ndim_array.py 	79 	25 	0 	68%
sympy/tensor/array/mutable_ndim_array.py 	3 	3 	0 	0%
sympy/tensor/array/ndim_array.py 	181 	68 	2 	62%
sympy/tensor/array/sparse_ndim_array.py 	97 	27 	0 	72%
sympy/tensor/index_methods.py 	128 	26 	2 	80%
sympy/tensor/indexed.py 	201 	55 	0 	73%
sympy/tensor/tensor.py 	1770 	371 	13 	79%
sympy/unify/__init__.py 	2 	0 	0 	100%
sympy/unify/core.py 	107 	7 	0 	93%
sympy/unify/rewrite.py 	18 	0 	0 	100%
sympy/unify/usympy.py 	56 	6 	0 	89%
sympy/utilities/__init__.py 	7 	7 	0 	0%
sympy/utilities/autowrap.py 	382 	133 	0 	65%
sympy/utilities/benchmarking.py 	127 	125 	0 	2%
sympy/utilities/codegen.py 	769 	57 	0 	93%
sympy/utilities/decorator.py 	80 	30 	0 	62%
sympy/utilities/enumerative.py 	321 	46 	0 	86%
sympy/utilities/exceptions.py 	28 	11 	0 	61%
sympy/utilities/iterables.py 	768 	138 	0 	82%
sympy/utilities/lambdify.py 	229 	55 	0 	76%
sympy/utilities/magic.py 	8 	2 	0 	75%
sympy/utilities/mathml/__init__.py 	20 	7 	0 	65%
sympy/utilities/memoization.py 	32 	21 	0 	34%
sympy/utilities/misc.py 	175 	116 	0 	34%
sympy/utilities/pkgdata.py 	18 	13 	0 	28%
sympy/utilities/pytest.py 	75 	13 	0 	83%
sympy/utilities/randtest.py 	77 	16 	0 	79%
sympy/utilities/runtests.py 	1082 	762 	0 	30%
sympy/utilities/source.py 	18 	9 	0 	50%
sympy/utilities/timeutils.py 	51 	40 	0 	22%
sympy/vector/__init__.py 	8 	0 	0 	100%
sympy/vector/basisdependent.py 	188 	19 	0 	90%
sympy/vector/coordsysrect.py 	165 	10 	0 	94%
sympy/vector/deloperator.py 	60 	2 	0 	97%
sympy/vector/dyadic.py 	113 	5 	0 	96%
sympy/vector/functions.py 	147 	10 	0 	93%
sympy/vector/orienters.py 	113 	11 	0 	90%
sympy/vector/point.py 	55 	4 	0 	93%
sympy/vector/scalar.py 	46 	3 	0 	93%
sympy/vector/vector.py 	180 	9 	0 	95%

coverage.py v4.2, created at 2016-12-24 01:00

mamueller committed:
linked the rcode function to the codegen docs

mamueller commented:
Thanks for the nice reception and the work that comes with the review.
I tried to follow the hints.

Cheers
Markus

2016-12-23 21:46 GMT+01:00 Jason K. Moore <notifications@github.com>:

> This is great! Thanks so much for this addition.
>
> A few things need to be fixed:
>
>    - Remove any trailing whitespace. You can run the test_code_quality.py
>    file to check. Also see the Travis failures.
>    - Can you add documentation? For example, make sure the API docs show
>    up like: http://docs.sympy.org/latest/modules/printing.html?
>    highlight=ccode#sympy.printing.ccode
>    <http://docs.sympy.org/latest/modules/printing.html?highlight=ccode#sympy.printing.ccode>.
>    And also would you mind adding a bit to here:
>    http://docs.sympy.org/latest/modules/codegen.html#code-
>    printers-sympy-printing
>    <http://docs.sympy.org/latest/modules/codegen.html#code-printers-sympy-printing>
>    - Can you run bin/coverage_report.py and report the results of your
>    test coverage here?
>
> Other than, that I think this is looking great!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/11973#issuecomment-269043862>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABETPTMKeohYgjVwykdXX8YVhe4hF9L8ks5rLDNBgaJpZM4LU6U5>
> .
>

moorepants commented:
The key one is this:

 sympy/printing/rcode.py 151 49 1 68%

Which is better than average for the code printers. That's fine for now I suppose, but it would be nice to increase the test coverage.

moorepants commented on sympy/printing/printer.py:
```diff
@@ -254,8 +254,9 @@ def _print(self, expr, *args, **kwargs):
-                    return getattr(self, printmethod)(expr, *args, **kwargs)
+                    res=getattr(self, printmethod)(expr, *args, **kwargs)
 
+                    return getattr(self, printmethod)(expr, *args, **kwargs)
```
This still needs to be fixed.

moorepants commented on sympy/printing/rcode.py:
```diff
@@ -0,0 +1,412 @@
+class RCodePrinter(CodePrinter):
+    """A printer to convert python expressions to strings of c code"""
+    printmethod = "_rcode"
+    language = "C"
```
These lines still have references to C.

mamueller committed:
Changed the few remaining forgotten references to C instead of R

mamueller commented:
I think that I addressed all issues by now.
In my python 3.5 environment also the ./bin/test and ./bin/doctest  yield no errors. So I wonder why the travis build has this remaining issue, (which I was not able to reproduce in my python 3.5.2 virtualenv)
I am not yet very familiar with the sympy pullrequest process but I suspect that the author's turn label could be removed now. 
Thanks for your time

bjodah commented:
I restarted the failed test on travis and it passes.
Thank you @mamueller for this contribution - it looks good, although I don't really know R, let's merge this now and iterate on it later if needed.

mamueller commented:
Thanks a lot.
I am happy that it worked out.
Cheers
Markus
